### PR TITLE
[Task 1471] Feedback Board Container Tests

### PR DESCRIPTION
**Pull request by msa2984**
**Task**:
https://dev.azure.com/csebostoncrew/Retrospectives-AzDO-Extension/_workitems/edit/1471

**Notes**:
- I had a lot of trouble getting a Feedback Board Container rendering test to work at first, so I created some tests for the child components hoping that would help expose where the issues were coming from. I figured these tests might as well be added in this PR since I already had them set up. 
- I moved a lot of the mocked components from the `feedbackItem` tests to a separate file as they could be reused by the `feedbackBoardContainer` tests.
- The `cssTestStub` file resolves some of the [errors](https://stackoverflow.com/questions/39418555/syntaxerror-with-jest-and-react-and-importing-css-files) while trying to import styling into a JEST test. 
- Mocking `window.matchMedia` in `setupTests.tsx` resolves an [error](https://github.com/akiran/react-slick/issues/742) with polyfills. Right now this error only occurs during the feedbackBoardContainer tests, but it seems like it is recommended to add this to overall setup for the tests as it is possible as tests for other components are added they may exhibit the same behavior.

image

### CODE REVIEW

File: RetrospectiveExtension.Frontend/tsconfig.test.json [hunk 1]
```diff
@@ -9,6 +9,7 @@
       "strict": true,
       "strictNullChecks": false,
       "target": "es5",
+      "downlevelIteration": true,
```
Comment (msa2984) [inline RetrospectiveExtension.Frontend/tsconfig.test.json:12]: https://www.typescriptlang.org/tsconfig#downlevelIteration - this is set in the regular `tsconfig.json`. This is required for preventing various issues with `Set`

File: RetrospectiveExtension.Frontend/package.json [hunk 2]
```diff
@@ -100,9 +100,11 @@
       "./components/__tests__/setupTests.tsx"
     ],
     "moduleNameMapper": {
-      "office-ui-fabric-react/lib/(.*)$": "office-ui-fabric-react/lib-commonjs/$1"
+      "office-ui-fabric-react/lib/(.*)$": "office-ui-fabric-react/lib-commonjs/$1",
+      "^.+\\.(css|less)$": "<rootDir>/components/__mocks__/cssTestStub.tsx"
     },
     "preset": "ts-jest/presets/js-with-babel",
+    "resetMocks": true,
```
Comment (msa2984) [inline RetrospectiveExtension.Frontend/package.json:106]: reset mock call count, etc. Required for expecting the appropriate number of calls for mocks which are reused across different test files.

File: RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx [hunk 3]
```diff
@@ -705,7 +705,7 @@ class FeedbackBoardContainer extends React.Component<FeedbackBoardContainerProps
     }
   }
 
-  private clickWorkflowStateCallback = (clickedElement: React.MouseEvent<HTMLElement>, newPhase: WorkflowPhase) => {
+  private clickWorkflowStateCallback = (clickedElement: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLDivElement>, newPhase: WorkflowPhase) => {
```
Comment (msa2984) [inline RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx:708]: In the props for the `WorkflowStage`, the definition for the `clickEventCallback` this is passed to is defined as so:
```
clickEventCallback: (clickedElement: React.MouseEvent<HTMLElement> | React.KeyboardEvent<HTMLDivElement>, newPhase: WorkflowPhase) => void;
```

without adding this or, JEST throws errors that there is a type assignment mismatch. The other option to resolve this issue would be to separate `clickEventCallback` into two separate calls, however I think that would significantly complicate the `WorkflowStage` component which is used by components other than this one.

File: RetrospectiveExtension.Frontend/components/__tests__/editableDocumentCardTitle.test.tsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+const mockedProps = mocked({
+  isDisabled: true,
+  isMultiline: false,
+  maxLength: 50,
+  title: "Mocked Title",
+  isChangeEventRequired: true,
+  onSave: jest.fn(() => {})
+});
+
+describe('Workflow Stage ', () => {
```
Comment (focusthenflex) [inline RetrospectiveExtension.Frontend/components/__tests__/editableDocumentCardTitle.test.tsx]:
```suggestion
describe('Editable Document Card Title ', () => {
```

**Review by polatengin — APPROVED** [review]
LGTM 👍

**Review by polatengin — APPROVED** [review]
LGTM

File: RetrospectiveExtension.Frontend/components/__tests__/feedbackBoardContainer.test.tsx [hunk 5]
```diff
@@ -1,15 +1,62 @@
...
 
-describe(`The Feedback Board Component`, () => {
-  it("Renders a Feedback Board Container Component", () => {
-    console.log(feedbackBoardContainerProps);
+describe('Feedback Board Container ', () => {
+  it('can be rendered without content.', () => {
+    const wrapper = shallow(<FeedbackBoardContainer {...feedbackBoardContainerProps} />);
+    expect(wrapper.children().dive().html()).toBe(
+      '<div class="ms-Spinner initialization-spinner root-41"><div class="ms-Spinner-circle ms-Spinner--large circle-42">' +
+      '</div><div class="ms-Spinner-label label-43">Loading...</div></div>');
```
Comment (focusthenflex) [inline RetrospectiveExtension.Frontend/components/__tests__/feedbackBoardContainer.test.tsx:60]: Is there an async method for waiting and testing when the data has been loaded successfully?

File: RetrospectiveExtension.Frontend/components/__tests__/workflowStage.test.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    expect(component.prop('aria-label')).toBe(`Not selected ${mockedProps.display} workflow stage`);
+  });
+
+  it('calls clickEventCallback when the Enter key is pressed.', () => {
+    mockedProps.isActive = true;
+    const wrapper = shallow(<WorkflowStage {...mockedProps} />);
+    const component = wrapper.children().dive();
+
+    component.simulate('keydown', { keyCode: 13 });
```
Comment (focusthenflex) [inline RetrospectiveExtension.Frontend/components/__tests__/workflowStage.test.tsx:35]: Yes! This gets around the annoying error that I was seeing on the feedback board container

File: RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx (hunk 3 above)
Comment (focusthenflex) [reply RetrospectiveExtension.Frontend/components/feedbackBoardContainer.tsx:708]: Yes! This gets around the annoying error that I was seeing on the feedback board container

**Review by focusthenflex — APPROVED** [review]
Approved with questions!

File: RetrospectiveExtension.Frontend/components/__tests__/feedbackBoardContainer.test.tsx (hunk 5 above)
Comment (msa2984) [reply RetrospectiveExtension.Frontend/components/__tests__/feedbackBoardContainer.test.tsx:60]: oh good question - so something we can do is use the `mount` command. The reason the data hasn't been loaded is that `shallow` doesn't really call the `componentDidMount` method which populates the data. TBH ... I had so many issues getting any test to work for this component. I think we would need to start mocking out all of the services in the `dal` directory in order to successfully call componentDidMount for these more complex components.

Comment (focusthenflex) [reply RetrospectiveExtension.Frontend/components/__tests__/feedbackBoardContainer.test.tsx:60]: Ah, ok, that makes sense, thank you!

**Pull request merged**